src/utils.c

```diff
@@ -115,25 +140,322 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+ */
+SEXP replace_dot_alias(SEXP x) {
+  if (isLanguage(x) && !isFunction(CAR(x))) {
+    if (CAR(x) == sym_bquote) // handling `.` inside bquote, #1912
+      return x;
+    if (CAR(x) == sym_dot)
+      SETCAR(x, sym_list);
+    for (SEXP t=x; t!=R_NilValue; t=CDR(t)) {
+      if (!isNull(CADR(t)))
+        SETCADR(t, replace_dot_alias(CADR(t)));
+    }
+  }
+  return x;
+}
+SEXP replace_dot_aliasR(SEXP x) {
+  return replace_dot_alias(MAYBE_REFERENCED(x) ? duplicate(x) : x);
+}
+
+/* funCols
+ * helper that calls lapply(DT, fun) to produce logical vector
+ */
+SEXP funCols(SEXP x, SEXP fun, SEXP rho) {
+  SEXP lfn = PROTECT(LCONS(sym_lapply, LCONS(x, LCONS(fun, R_NilValue))));
+  SEXP lst = PROTECT(eval(lfn, rho));
+  SEXP log = PROTECT(allocVector(LGLSXP, length(x)));
+  int *logp = LOGICAL(log);
+  for (int i=0; i<length(lst); ++i) {
+    SEXP lsti = VECTOR_ELT(lst, i);
+    if (length(lsti)!=1 || !isLogical(lsti) || LOGICAL(lsti)[0]==NA_LOGICAL)
+      error("When argument for columns selection is a function, it is applied to each column; the output of this function must be a non-missing boolean scalar signalling inclusion/exclusion of the column. However, these conditions were not met at least for: %s", CHAR(STRING_ELT(getAttrib(x, R_NamesSymbol), i)));
+    logp[i] = LOGICAL(lsti)[0];
+  }
+  UNPROTECT(3);
+  return log;
+}
+
+/* exprCols
+ * turns NSE expression into columns int
+ * helper for j and .SDcols
+ * handles v1, "v1", v1:v3, !"v1", -"v1", paste0("v",1:3)
+ * does not evaluate only for `symbol:symbol` expression, or `symbol` when "symbol" is existing column
+ * with is updated by reference thus should be copied before being passed to this function
+ */
+SEXP exprCols(SEXP x, SEXP expr, SEXP mode, SEXP with, SEXP rho) {
+  int protecti=0;
+  if (!isString(mode) || length(mode)!=1 || STRING_ELT(mode, 0)==NA_STRING)
+    error(_("internal error: mode argument has to be character, non-NA of length 1")); // # nocov
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=NULL
+  if (!isNewList(x))
+    error(_("'x' argument must be data.table"));
+  bool mode_j=false, mode_sd=false;
+  if (strcmp(CHAR(STRING_ELT(mode, 0)), "j")==0)
+    mode_j = true;
+  else if (strcmp(CHAR(STRING_ELT(mode, 0)), ".SDcols")==0)
+    mode_sd = true;
+  else
+    error(_("Internal error: invalid 'mode' argument in exprCols function, should have been caught before. please report to data.table issue tracker.")); // # nocov
+  bool inverse = isLanguage(expr) && (CAR(expr)==sym_bang || (CAR(expr)==sym_minus && length(expr)==2)); // length(expr)==2, otherwise V3-V2 will pick it up
+  if (inverse)
+    expr = CADR(expr);
+  bool peeled = false;
+  SEXP sym_parenthesis = install("(");
+  while (isLanguage(expr) && CAR(expr)==sym_parenthesis) {
+    expr = CADR(expr);
+    peeled = true;
+  }
+  SEXP sym_brackets = install("{");
+  while (isLanguage(expr) && CAR(expr)==sym_brackets && length(expr)==2) {
```

Comment:
@jangorecki Is ```isLanguage``` equivalent to ```is.call()```? Also, why do the brackets need the ```length(expr) == 2)```? Is that to only peel one layer? Also, I'm not sure for etiquette on closed PRs or if you even receive notifications.

Reply:
here's the source for `is.call`

https://github.com/wch/r-source/blob/trunk/src/main/coerce.c#L2072-L2073

```
    case 300:		/* is.call */
	LOGICAL0(ans)[0] = TYPEOF(CAR(args)) == LANGSXP;
```

Reply:
_almost_ identical to `isLanguage`:

https://github.com/wch/r-source/blob/trunk/src/include/Rinlinedfuns.h#L903-L906

```
INLINE_FUN Rboolean isLanguage(SEXP s)
{
    return (s == R_NilValue || TYPEOF(s) == LANGSXP);
}
```

Reply:
So it appears that for how we set up our functions, ```TYPEOF(s) == LANGSXP``` would be equivalent of ```is.call()``` since I believe ```args``` is normally a list for the C arguments. I am asking because I am interested in a more direct ```%iscall%```. Thanks!

Reply:
only `{` needs length check because `length(quote({1; 2}))` should not be peeled. `(` will be always length 2.